reference/JSDoc-Type-References.md

```diff
@@ -0,0 +1,242 @@
+
+But what if you treat `i` as a type?
+
+```js
+/** @type {i} */
+var j
+```
+
+The compiler just reuses the type of the value declaration: `number`.
+This is equivalent to inserting the `typeof` type operator to get
+`typeof i`.
+The effect is to make types more Javascripty, allowing you to specify
+objects as example types without having to learn about the difference
+between type and value:
+
+``` js
+/**
+ * @param {number} n
+ * @param {number} m
+ * @returns {-1 | 0 | 1}
+ */
+function exampleCompare(n, m) {
+}
+
+/** @param {exampleCompare} f */
+function sort(f, l) {
+  // ...
+}
+```
+
+This also works with objects, whether anonymous or an instance of a class:
+
+``` js
+const initial = {
+  frabjous: true,
+  beamish: true,
+  callooh: "callay",
+}
+
+/** @param {initial} options */
+function setup(options) {
+}
+```
+
+Relying on this fallback for complex types does produce a simpler
+type, but makes the equivalent Typescript really confusing to produce:
+
+``` js
+import { options } from './initial'
+/**
+ * @param {keyof options} k
+ * @param {options[keyof options]} v
+ */
+function demo(k, v) {
+    options[k] = v
+}
+```
+
+is equivalent to the Typescript:
+
+```ts
+import { options } from './initial'
+function demo(k: keyof typeof options, v: typeof options[keyof typeof options]) {
+}
+```
+
+Knowing where to insert the missing `typeof` operators requires more
+experience with complex type than most people can be expected to have.
+
+### Where To Find The Code ###
+
+In src/compiler/checker.ts, `getTypeFromTypeReference` is a good
+starting point. The two fallbacks are in `resolveEntityName` and
+`getTypeReferenceType`, respectively.
+
+## Expando ##
+
+Expando declarations are those that start with an initial declaration
+and add properties to it, like so:
+
+``` js
+const ns = {}
+ns.x = 1
+ns.f = function () {
+}
+ns.s = 'otra'
+```
+
+In Javascript code, the compiler supports this pattern for
+construction of objects. It also supports it for functions and classes.
+
+Syntactically, expando objects are of 3 kinds:
+
+1. Functions
+2. Classes
+3. Empty objects
+
+Each expando assignment is treated as a separate namespace
+declaration, which then merges with all the other declarations of the
+object. That is, the following code declares one variable with three
+declarations; one value declaration and two namespace declarations.
+The last two lines also each declare one property with a value declaration.
+
+```js
+function ichthyosaur() { }
+ichthyosaur.shonisaurus = function() { }
+ichthyosaur.shastasaurus = function() { }
+```
+
+TODO: Explain aliasing.
+
+### Where To Find The Code ###
+
+Object and class expandos are checked just like Typescript. Only their
+binding differs: in src/compiler/binder.ts,
+`bindSpecialPropertyAssignment` delegates to `bindPropertyAssignment`.
+
+## CommonJS ##
+
+Like expando, except that you are often directly making entries in the
+module's symbol table instead.
+
+TODO: Explain aliasing.
```

Comment:
I'm not sure whether this TODO is repeated or maybe it is about some other kind of aliasing (since I don't know how commonjs is related here, isn't it just module-level thing?).

Reply:
The compiler tries even harder to figure out patterns of aliasing for `module.exports`, which means that this section will expand on the previous one.